### PR TITLE
fix(slack/channels):Add pagination to channels

### LINKED ISSUE #83: Feature: Slack notifications
Original requirement is here: https://app.clubhouse.io/flagsmith/story/182/slack-integration
Code is in a feature branch in Gitlab (merge request 194) which can be migrated over to Github and using as a starting reference.

### PR DESCRIPTION
**Pull request by gagantrivedi**

### CODE REVIEW

File: api/integrations/slack/views.py [hunk 1]
```diff
@@ -19,40 +21,45 @@
...
 from .permissions import OauthInitPermission
 
 signer = TimestampSigner()
 
 
+class SlackGetChannelsViewSet(GenericViewSet):
+    serializer_class = SlackChannelListSerializer
+    pagination_class = ChannelListPagination
+
+    def get_queryset(self) -> SlackWrapper:
```
Comment (matthewelwell) [inline api/integrations/slack/views.py]: Hmm... I'm not sure about the fact that we're returning a `SlackWrapper` object here? The docs for `get_queryset` state: 

```
        Get the list of items for this view.
        This must be an iterable, and may be a queryset.
        Defaults to using `self.queryset`.
```

File: api/integrations/slack/views.py [hunk 2]
```diff
@@ -19,40 +21,45 @@
...
+        try:
+            config = SlackConfiguration.objects.get(project=environment.project)
+        except ObjectDoesNotExist as e:
+            raise SlackConfigurationDoesNotExist() from e
+        slack_wrapper = SlackWrapper(api_token=config.api_token)
+
+        return slack_wrapper
+
+    def list(self, request, *args, **kwargs):
+        queryset = self.filter_queryset(self.get_queryset())
```
Comment (matthewelwell) [inline api/integrations/slack/views.py]: I think adding `filter_queryset` is probably a bad idea here. The only filter backend we have is the `DjangoFilterBackend` which will likely throw some exception if invoked (which I think would only happen if someone accidentally added a `filterset_fields` attribute to this class but still).

File: api/integrations/slack/views.py (hunk 1 above)
Comment (gagantrivedi) [reply api/integrations/slack/views.py]: get_queryset -> get_slack_wrapper ?

File: api/integrations/slack/views.py [hunk 3]
```diff
@@ -19,40 +21,44 @@
...
 from .permissions import OauthInitPermission
 
 signer = TimestampSigner()
 
 
+class SlackGetChannelsViewSet(GenericViewSet):
+    serializer_class = SlackChannelListSerializer
+    pagination_class = ChannelListPagination
+
+    def get_api_token(self) -> SlackWrapper:
```
Comment (matthewelwell) [inline api/integrations/slack/views.py]: I don't think this typehint is correct anymore?

File: api/integrations/slack/pagination.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+from rest_framework.response import Response
+
+
+class ChannelListPagination(CursorPagination):
+    page_size_query_param = "page_size"
+
+    def paginate_queryset(self, queryset, request, view=None):
+        page_size = self.get_page_size(request)
+        cursor = request.GET.get("cursor", None)
+        response = queryset.get_channels_data(limit=page_size, cursor=cursor)
```
Comment (matthewelwell) [inline api/integrations/slack/pagination.py]: I do think this use of the pagination class is quite neat but I'm not sure about our use of `queryset` here. It's a bit misleading to have a queryset that is not iterable. I think we should rethink this slightly. Maybe `get_channels_data` should return some kind of iterable?

Comment (matthewelwell) [reply api/integrations/slack/pagination.py]: The more I think about this, I don't think we're approaching it correctly. We're not _Really_ doing the pagination. That's why it feels a little bit like we're shoehorning our logic into a django pagination class. 

Essentially all we're doing is proxying a request and doing some serialization. 

I think this code (views, serializers, pagination) would be neater if we just dropped all of the pagination stuff and just had an endpoint that accepts some get parameters and returns some serialized data. 

Something along these lines?

```
# serializers.py

class GetChannelsQuerySerializer(serializers.Serializer):
    cursor = serializers.CharField()
    limit = serializers.IntegerField()

# views.py

class SlackChannelsViewSet(GenericViewSet):
    def list(self, request, *args, **kwargs):
        query_serializer = GetChannelsQuerySerializer(data=request.data)
        query_serializer.is_valid(raise_exception=True)
        
        api_token = self.get_api_token()
        slack_wrapper = SlackWrapper(api_token=api_token)
        
        channels_data = slack_wrapper.get_channels_data(**query_serializer.data)
        return self.get_serializer(channels_data)
```

Comment (gagantrivedi) [reply api/integrations/slack/pagination.py]: yeah, that makes sense

**Review by matthewelwell — APPROVED** [review]

**Pull request merged**